p256/fiat_p256.mli

```diff
@@ -1,61 +1,31 @@
-(** Convert from cstruct. Refuses the point at infinity.
-    The format is the uncompressed format described in
-    SEC1, section 2.3.4, that is to say:
-
-    - the point at infinity is the single byte "00".
-    - for a point (x, y) not at infinity, the format is:
-      - the byte "04"
-      - x serialized in big endian format, padded to 32 bytes.
-      - y serialized in big endian format, padded to 32 bytes.
-
-    @see <http://www.secg.org/sec1-v2.pdf>
-*)
-
-val point_of_hex : Hex.t -> (point, point_error) result
-(** Convert a point from hex. See [point_of_cs]. *)
-
-val point_to_cs : point -> Cstruct.t
-(** Convert a point to a cstruct. See [point_of_cs] for the format. *)
-
-(** A scalar value. *)
-type scalar
+val pp_error : Format.formatter -> error -> unit
+(** Pretty printer for public key parsing errors *)
 
-(** The type for scalar parsing errors. *)
-type scalar_error =
-  [ `Invalid_length
-  | `Invalid_range ]
+(** Type for P256 private keys *)
+type secret
 
-val pp_scalar_error : Format.formatter -> scalar_error -> unit
-(** Pretty printer for scalar parsing errors *)
+val gen_key : rng:(int -> Cstruct.t) -> secret * Cstruct.t
+(** [gen_key ~rng] generates a private and a public key for Ephemeral Diffie-Hellman
+    over P256. The returned key pair MUST only be used for a single key exchange.
 
-val scalar_of_cs : Cstruct.t -> (scalar, scalar_error) result
-(** Read data from a cstruct.
-    It should be 32 bytes long, in big endian format. Returns an error when the
-    number is zero, or if it is larger than or equal to the group order. *)
+    [rng] is the function used to repeteadly generate a private key until a valid candidate
+    is obtained. [rng]'s int parameter is the size of the [Cstruct.t] to generate.
+    If [rng] returns an invalid length buffer, [Failure _] is raised.
 
-val scalar_of_hex : Hex.t -> (scalar, scalar_error) result
-(** Like [scalar_of_cs] but read from hex data. *)
+    The generated private key is checked to be greater than zero and lower than the group
+    order meaning the public key cannot be the point at inifinity. *)
 
-val dh : scalar:scalar -> point:point -> Cstruct.t
-(** Perform Diffie-Hellman key exchange. This returns the x component of the
-    scalar multiplication of [point] and [scalar]. *)
+val key_exchange : secret -> Cstruct.t -> (Cstruct.t, error) result
+(** [key_exchange ~private_key received_public_key] performs Diffie-Hellman key exchange
+    using your private key and the other party's public key. Returns the shared secret
+    or an error if the received public is invalid or is the point at infinity.
 
-val public : scalar -> point
-(** Compute the public key corresponding to a given private key. Internally,
-    this multiplies the generator by the scalar.
-    Given the invariant on [scalar], the result can't be the point at infinity. *)
+    @see <http://www.secg.org/sec1-v2.pdf> for public key encoding format. *)
```

Comment:
here as well, please mention truncation (i suspect the shared secret is not truncated).

what is an `invalid` public (i suspect bad encoding, not on curve)?